llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -581,30 +581,30 @@ class InnerLoopVectorizer {
   virtual void printDebugTracesAtEnd() {}
 
   /// The original loop.
-  Loop *OrigLoop;
+  Loop *const OrigLoop;
 
   /// A wrapper around ScalarEvolution used to add runtime SCEV checks. Applies
   /// dynamic knowledge to simplify SCEV expressions and converts them to a
   /// more usable form.
   PredicatedScalarEvolution &PSE;
 
   /// Loop Info.
-  LoopInfo *LI;
+  LoopInfo *const LI;
```

Comment:
no east const

Reply:
const LoopInfo *

Reply:
My change says "`LI` cannot be mutated".
Your change would say "`*LI` cannot be mutated" and it does not compile:

```diff
diff --git a/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp b/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
index ca89d7676677..d4cf5046b722 100644
--- a/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
+++ b/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
@@ -589,7 +589,7 @@ protected:
   PredicatedScalarEvolution &PSE;

   /// Loop Info.
-  LoopInfo *const LI;
+  const LoopInfo *LI;

   /// Dominator Tree.
   DominatorTree *const DT;
```

```
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp: In member function ‘void llvm::InnerLoopVectorizer::emitIterationCountCheck(llvm::BasicBlock*)’:
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2498:67: error: invalid conversion from ‘const llvm::LoopInfo*’ to ‘llvm::LoopInfo*’ [-fpermissive]
 2498 |       SplitBlock(TCCheckBlock, TCCheckBlock->getTerminator(), DT, LI, nullptr,
      |                                                                   ^~
      |                                                                   |
      |                                                                   const llvm::LoopInfo*
In file included from /home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:135:
/home/p.fusik/upstream/llvm-project/llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:288:34: note:   initializing argument 4 of ‘llvm::BasicBlock* llvm::SplitBlock(llvm::BasicBlock*, llvm::Instruction*, llvm::DominatorTree*, llvm::LoopInfo*, llvm::MemorySSAUpdater*, const llvm::Twine&, bool)’
  288 |                        LoopInfo *LI = nullptr,
      |                        ~~~~~~~~~~^~~~~~~~~~~~
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp: In member function ‘void llvm::InnerLoopVectorizer::createVectorLoopSkeleton(llvm::StringRef)’:
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2577:18: error: invalid conversion from ‘const llvm::LoopInfo*’ to ‘llvm::LoopInfo*’ [-fpermissive]
 2577 |                  LI, nullptr, Twine(Prefix) + "middle.block");
      |                  ^~
      |                  |
      |                  const llvm::LoopInfo*
In file included from /home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:135:
/home/p.fusik/upstream/llvm-project/llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:288:34: note:   initializing argument 4 of ‘llvm::BasicBlock* llvm::SplitBlock(llvm::BasicBlock*, llvm::Instruction*, llvm::DominatorTree*, llvm::LoopInfo*, llvm::MemorySSAUpdater*, const llvm::Twine&, bool)’
  288 |                        LoopInfo *LI = nullptr,
      |                        ~~~~~~~~~~^~~~~~~~~~~~
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:2579:73: error: invalid conversion from ‘const llvm::LoopInfo*’ to ‘llvm::LoopInfo*’ [-fpermissive]
 2579 |       SplitBlock(LoopMiddleBlock, LoopMiddleBlock->getTerminator(), DT, LI,
      |                                                                         ^~
      |                                                                         |
      |                                                                         const llvm::LoopInfo*
In file included from /home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:135:
/home/p.fusik/upstream/llvm-project/llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:288:34: note:   initializing argument 4 of ‘llvm::BasicBlock* llvm::SplitBlock(llvm::BasicBlock*, llvm::Instruction*, llvm::DominatorTree*, llvm::LoopInfo*, llvm::MemorySSAUpdater*, const llvm::Twine&, bool)’
  288 |                        LoopInfo *LI = nullptr,
      |                        ~~~~~~~~~~^~~~~~~~~~~~
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp: In member function ‘llvm::BasicBlock* llvm::EpilogueVectorizerMainLoop::emitIterationCountCheck(llvm::BasicBlock*, bool)’:
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7882:40: error: invalid conversion from ‘const llvm::LoopInfo*’ to ‘llvm::LoopInfo*’ [-fpermissive]
 7882 |                                    DT, LI, nullptr, "vector.ph");
      |                                        ^~
      |                                        |
      |                                        const llvm::LoopInfo*
In file included from /home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:135:
/home/p.fusik/upstream/llvm-project/llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:288:34: note:   initializing argument 4 of ‘llvm::BasicBlock* llvm::SplitBlock(llvm::BasicBlock*, llvm::Instruction*, llvm::DominatorTree*, llvm::LoopInfo*, llvm::MemorySSAUpdater*, const llvm::Twine&, bool)’
  288 |                        LoopInfo *LI = nullptr,
      |                        ~~~~~~~~~~^~~~~~~~~~~~
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp: In member function ‘virtual std::pair<llvm::BasicBlock*, llvm::Value*> llvm::EpilogueVectorizerEpilogueLoop::createEpilogueVectorizedLoopSkeleton(const SCEV2ValueTy&)’:
/home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7923:73: error: invalid conversion from ‘const llvm::LoopInfo*’ to ‘llvm::LoopInfo*’ [-fpermissive]
 7923 |       SplitBlock(LoopVectorPreHeader, LoopVectorPreHeader->begin(), DT, LI,
      |                                                                         ^~
      |                                                                         |
      |                                                                         const llvm::LoopInfo*
In file included from /home/p.fusik/upstream/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:135:
/home/p.fusik/upstream/llvm-project/llvm/include/llvm/Transforms/Utils/BasicBlockUtils.h:284:34: note:   initializing argument 4 of ‘llvm::BasicBlock* llvm::SplitBlock(llvm::BasicBlock*, llvm::BasicBlock::iterator, llvm::DominatorTree*, llvm::LoopInfo*, llvm::MemorySSAUpdater*, const llvm::Twine&, bool)’
  284 |                        LoopInfo *LI = nullptr,
      |                        ~~~~~~~~~~^~~~~~~~~~~~
```

Reply:
So basically this isn't const at all

Reply:
What do you mean? It literally says `const`.

Reply:
I think this change is saying that the member variable of `InnerLoopVectorizer` which holds a pointer to LoopInfo doesn't change after `InnerLoopVectorizer` is constructed? But InnerLoopVectorizer can make changes to the LoopInfo object it points to?

Reply:
I'm not a fan of putting `const` everywhere, for example on local variables and parameters.
But I see value in saying that a member variable is not mutated after construction. I'm surprised that you weren't aware of this meaning of `const` !
Also, to be clear, it is not my goal to add `const`s here and there, but to understand the code.

Reply:
Here's why `const` member variables are useful: https://godbolt.org/z/qKooahxYx - let me know if that's clear!

Reply:
Yes, I know what it does. But is there a big picture here? Are you trying to add const all over LLVM? If so that needs an RFC. If not, well, why are these files so special that it's worth doing? We generally like to avoid churn unless there's a good reason for it.

Reply:
> Are you trying to add const all over LLVM?

No.

> If not, well, why are these files so special that it's worth doing?

These are the files I was reading to make some changes later. As I'm reading, I'm fixing small issues such as typos, missing `const` etc. Is that bad?